### ISSUE TITLE
Using custom Django hashers, and being able to edit the hashers constant on the Django settings file (User migration from other E-com platforms)

### ISSUE THREAD

**Issue by Dieg0x17** [saleor/saleor#10419]
Using custom Django hashers, and being able to edit the hashers constant on the Django settings file (User migration from other E-com platforms). This is a common problem when users want to migrate from another E-com platform with hashing algorithms not supported by default on the Django framework. Django support extending this feature with custom objects.

We are facing some limitations if we want to use the Saleor cloud or we want to avoid editing the saleor-core to avoid maintaining a fork.

- Being able to edit the hasher list constant on the Django settings file (Using env variables to set the list, could be a solution).
- Being able to add custom packages with the custom Django Hashers objects not supported by default on Django to the default project dependencies.

**Comment by NyanKiyoshi** [comment]
Hi @Dieg0x17,

What kind of hashes are you trying to import into Saleor? Software name and version would be useful to determine how the support could be implemented

**Comment by Dieg0x17** [comment]
My particular case is this hashing algorithm, other people could face the same problem with different algorithms and just covering my case is not the best solution, a general solution that can cover those could be better.

```
import base64
import hashlib
from collections import OrderedDict

from django.utils.crypto import (
    constant_time_compare, )
from django.utils.encoding import force_bytes
from django.utils.translation import gettext_noop as _
from django.contrib.auth.hashers import BasePasswordHasher, mask_hash


class UnsaltedBase64SHA512PasswordHasher(BasePasswordHasher):
    """
    Synfony sha512 encoder
    """
    algorithm = "unsalted_b64_sha512"

    def salt(self):
        return ''

    def encode(self, password, salt='', iters=5000):
        assert salt == ''
        digest = hashlib.sha512(force_bytes(password, encoding="ISO-8859-1")).digest()  # ISO-8859-1
        for i in range(iters - 1):
            digest = hashlib.sha512(digest + force_bytes(password, encoding="ISO-8859-1")).digest()
        return 'b64sha512$$%s' % base64.b64encode(digest).decode("ISO-8859-1")

    def verify(self, password, encoded):
        encoded_2 = self.encode(password, '')
        return constant_time_compare(encoded, encoded_2)

    def safe_summary(self, encoded):
        assert encoded.startswith('b64sha512$$')
        hash = encoded[11:]
        return OrderedDict([
            (_('algorithm'), self.algorithm),
            (_('hash'), mask_hash(hash)),
        ])

    def harden_runtime(self, password, encoded):
        pass
```

**Comment by NyanKiyoshi** [comment]
Hi @Dieg0x17, it looks like the base64 encoding is custom. We could add a general support for SHA512 encoded password directly to Saleor and you could adapt the users hashes from your database dump so they are compatible.

The following sample Python code would allow to convert from Base64 SHA512 hashes to plain-text SHA512 hashes:
```py
import base64

# Base64 encoded SHA512 hash bytes from 'admin'
my_hash = "x61Ey612Kl2gpFL56FT9weDnpSo4AV8j8+qx2AuTHdRyY036xxzTTrw10Wq3+4qQyB+XURPWx1ONxp3Y3pB37A=="

# Decode hash to bytes
sha512_bytes: bytes = base64.b64decode(my_hash)

# c7ad44cbad762a5da0a452f9e854fdc1e0e7a52a38015f23f3eab1d80b931dd472634dfac71cd34ebc35d16ab7fb8a90c81f975113d6c7538dc69dd8de9077ec
sha512_bytes.hex()

**Comment by Dieg0x17** [comment]
I need to transform the password strings anyways, that could be a valid solution for my particular case.
I thought sha512 was not implemented by default in Django hashers only sha1 and sha256.

Could you please give me an example of the hash token prefix that the hasher will use to detect the algo?

thanks

**Comment by Dieg0x17** [comment]
Do you think we can have problems with the python utf-8 encoding? 
The hashes were encoded using 1 byte Latin-1 (PHP default encoding).

**Comment by NyanKiyoshi** [comment]
Hi @Dieg0x17,

>  I thought sha512 was not implemented by default in Django hashers only sha1 and sha256.

SHA512 support is not implemented by Django, we would add it to the Saleor code based on the [sha1 implementation of Django](https://github.com/django/django/blob/3b79dab19a2300a4884a3d81baa6c7c1f2dee059/django/contrib/auth/hashers.py#L721). The prefix would be `unsalted_sha512`, for example:
```
unsalted_sha512$$c7ad44cbad762a5da0a452f9e854fdc1e0e7a52a38015f23f3eab1d80b931dd472634dfac71cd34ebc35d16ab7fb8a90c81f975113d6c7538dc69dd8de9077ec
```

> Do you think we can have problems with the python utf-8 encoding?
> The hashes were encoded using 1 byte Latin-1 (PHP default encoding).

It shouldn't cause issues.

**Comment by NyanKiyoshi** [comment]
There is no round count support for standard SHA512 hashes (i.e. when not using bcrypt)

This would be implemented the same way as Django does for plain SHA1 hashes: https://github.com/django/django/blob/3b79dab19a2300a4884a3d81baa6c7c1f2dee059/django/contrib/auth/hashers.py#L721

Once SHA512 support is implemented into Saleor the users will automatically be upgraded to the Django default hashing algorithm (PBKDF2) upon login

**Comment by Dieg0x17** [comment]
Our migrated hashes have 5000 rounds. It could be useful having a general solution and be able to add new hasher objects with custom implementations, to achieve a general solution instead of fixing a particular algorithm case.

It will be fine if we can just load from an environment variable extra elements for the settings PASSWORD_HASHERS constant list and add the custom hasher module to the project dependencies if we can do that without maintaining a fork.

**Comment by Dieg0x17** [comment]
Thanks

**Comment by Dieg0x17** [comment]
@CossackDex

**Comment by patrys** [comment]
Closing as fixed by #13239
